apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/profile/ProfileThread.java

```diff
@@ -0,0 +1,296 @@
+            } catch (InterruptedException e) {
+                continue;
+            }
+
+            try {
+                profiling(taskExecutionContext);
+            } catch (InterruptedException e) {
+                // ignore interrupted
+                continue;
+            } catch (Exception e) {
+                logger.error(e, "Profiling task fail. taskId:{}", taskExecutionContext.getTask().getTaskId());
+            } finally {
+                // finally stop current profiling task, tell execution service task has stop
+                profileTaskExecutionService.stopCurrentProfileTask(taskExecutionContext);
+            }
+        }
+    }
+
+    /**
+     * notify have new task need to process
+     */
+    void processNewProfileTask(ProfileTaskExecutionContext context) {
+        executionContextListener.add(context);
+    }
+
+    /**
+     * check have available slot to profile and add it
+     * @param segment
+     * @return
+     */
+    public ProfilingSegmentContext checkAndAddSegmentContext(TraceSegment segment, ProfileTaskExecutionContext taskExecutionContext) {
+        // check has available slot
+        AtomicInteger currentProfilingCount = taskExecutionContext.getCurrentProfilingCount();
+        final int usingSlotCount = currentProfilingCount.get();
+        if (usingSlotCount >= Config.Profile.MAX_PARALLEL) {
+            return null;
+        }
+
+        // try to occupy slot
+        if (!currentProfilingCount.compareAndSet(usingSlotCount, usingSlotCount + 1)) {
+            return null;
+        }
+
+        ProfilingSegmentContext[] profilingSegmentSlot = taskExecutionContext.getProfilingSegmentSlot();
+        final ProfilingSegmentContext segmentContext = new ProfilingSegmentContext(segment, Thread.currentThread(), taskExecutionContext);
+        for (int slot = 0; slot < profilingSegmentSlot.length; slot++) {
+            if (profilingSegmentSlot[slot] == null) {
+                profilingSegmentSlot[slot] = segmentContext;
+                break;
+            }
+        }
+        return segmentContext;
+    }
+
+    /**
+     * find segment and clear on slot
+     *
+     * @param segment
+     */
+    public void stopSegmentProfile(TraceSegment segment) {
+        ProfileTaskExecutionContext currentExecutionContext = profileTaskExecutionService.getCurrentTaskExecutionContext();
+        if (currentExecutionContext == null) {
+            return;
+        }
+
+        // find current segment and clear it
+        boolean find = false;
+        ProfilingSegmentContext[] profilingSegmentSlot = currentExecutionContext.getProfilingSegmentSlot();
+        for (int slot = 0; slot < profilingSegmentSlot.length; slot++) {
+            ProfilingSegmentContext currentSlotSegment = profilingSegmentSlot[slot];
+            if (currentSlotSegment != null && Objects.equal(profilingSegmentSlot[slot].getSegment().getTraceSegmentId(), segment.getTraceSegmentId())) {
```

Comment:
You should be able to use `segment#equal`, right? Also, you should begin to move the logic method into the entity. 
This should be `profilingSegmentSlot[slot]`.`matchWith(segment`.